File: leetcode/problems/problem-specific/242.valid-anagram/242.valid-anagram.go
package main

/*
 * @lc app=leetcode id=242 lang=golang
 *
 * [242] Valid Anagram
 */

// @lc code=start

// * Hash Table Solution | O(n) Time | O(1) Space
// * n -> Number of letters in s/t

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	letterMappings := [256]int{}
	for _, v := range t {
		letterMappings[v]++
	}

	for _, v := range s {
		letterMappings[v]--
		if letterMappings[v] < 0 {
			return false
		}
	}

	return true
}

// @lc code=end

File: leetcode/problems/problem-specific/242.valid-anagram/242.valid-anagram.py
#
# @lc app=leetcode id=242 lang=python3
#
# [242] Valid Anagram
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(n) Space
# * n -> The length of s or t string


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        t_counter = collections.Counter(t)
        for char in s:
            t_counter[char] -= 1
            if t_counter[char] < 0:
                return False

        return True


# @lc code=end
